subsys/modem/backends/modem_backend_stat.c

```diff
@@ -0,0 +1,192 @@
+{
+	struct modem_backend_stat *next_stat;
+
+	K_SPINLOCK(&stat_lock) {
+		next_stat = stat_from_node(sys_slist_peek_next(&stat->node));
+	}
+
+	return next_stat;
+}
+
+static bool stat_is_used(uint32_t capacity)
+{
+	return capacity > 0;
+}
+
+static bool stat_is_valid(uint32_t max, uint32_t capacity)
+{
+	return max <= capacity;
+}
+
+static uint8_t percent_used(uint32_t max, uint32_t cap)
+{
+	uint64_t percent;
+
+	if (max == 0) {
+		return 0;
+	}
+
+	if (max == cap) {
+		return 100;
+	}
+
+	percent = 100;
+	percent <<= 16;
+	percent /= cap;
+	percent *= max;
+	percent >>= 16;
+
+	return (uint8_t)percent;
+}
+
+static void log_used_stat(const char *name, const char *dir, uint32_t max, uint32_t cap)
+{
+	LOG_PRINTK("backend stat: %s: %s: used: %u of %u (%u%%)\n", name, dir, max, cap,
+		   percent_used(max, cap));
+}
+
+static void log_invalid_stat(const char *name, const char *dir, uint32_t max, uint32_t cap)
+{
+	LOG_ERR("backend stat: %s: %s: exceeds capacity: %u > %u\n", name, dir, max, cap);
+}
+
+static void process_stat(const char *name, const char *dir, uint32_t max, uint32_t cap)
+{
+	if (!stat_is_used(cap)) {
+		return;
+	}
+
+	if (stat_is_valid(max, cap)) {
+		log_used_stat(name, dir, max, cap);
+	} else {
+		log_invalid_stat(name, dir, max, cap);
+	}
+}
+
+static void process_transmit_stat(const struct modem_backend_stat *stat, uint32_t max)
+{
+	process_stat(stat->name, "transmit", max, stat->transmit_buf_capacity);
+}
+
+static void process_receive_stat(const struct modem_backend_stat *stat, uint32_t max)
+{
+	process_stat(stat->name, "receive", max, stat->receive_buf_capacity);
+}
+
+static void get_and_clear_stat_max(struct modem_backend_stat *stat, uint32_t *transmit_buf_max,
+			 uint32_t *receive_buf_max)
+{
+	K_SPINLOCK(&stat_lock) {
+		*transmit_buf_max = stat->transmit_buf_max;
+		*receive_buf_max = stat->receive_buf_max;
+		stat->transmit_buf_max = 0;
+		stat->receive_buf_max = 0;
+	}
+}
+
+static void update_stat_max(struct modem_backend_stat *stat, uint32_t transmit_buf_size,
+			    uint32_t receive_buf_size)
+{
+	K_SPINLOCK(&stat_lock) {
+		if (stat->transmit_buf_max < transmit_buf_size) {
+			stat->transmit_buf_max = transmit_buf_size;
+		}
+
+		if (stat->receive_buf_max < receive_buf_size) {
+			stat->receive_buf_max = receive_buf_size;
+		}
+	}
+}
+
+static void stat_work_handler(struct k_work *work)
+{
+	struct modem_backend_stat *stat;
+	uint32_t transmit_buf_max;
+	uint32_t receive_buf_max;
+
+	ARG_UNUSED(work);
+
+	stat = get_first_stat();
+
+	while (stat != NULL) {
+		get_and_clear_stat_max(stat, &transmit_buf_max, &receive_buf_max);
+		process_transmit_stat(stat, transmit_buf_max);
+		process_receive_stat(stat, receive_buf_max);
+		stat = get_next_stat(stat);
+	}
+
+	k_work_schedule(&stat_work, STAT_WORK_DELAY);
+}
+
+static int z_modem_backend_stat_init(void)
+{
+	sys_slist_init(&stat_list);
+	k_work_init_delayable(&stat_work, stat_work_handler);
+	k_work_schedule(&stat_work, STAT_WORK_DELAY);
```

Comment:
We have similar feature in networking (printing statistics at interval), it was a good thing at the beginning but has not been very useful lately. We placed the background printing behind a Kconfig option because it was just very annoying to get prints when you do not need them. Instead there is now `net stats` net-shell command which prints the statistics when needed. So with this mind I suggest to place this behind an option and also add a way to manually see the statistics.

Reply:
I will mimic the way it is done for net stats :)